### PR TITLE
Convert to mci container

### PR DESCRIPTION
**Pull request by benroy73**
Here's the initial starting point for an example flask app that is ready to deploy to the MCI platform. It's running here https://tools-cloud-dev.s.uw.edu/example-flask-app/

Please take a look at the README file and try it out. It runs locally the same as in the cloud.

If you see any glaring issues or something that is confusing or doesn't work let me know and I'll fix it here. Once that's out of the way I'll merge it to master and we can start improving/simplifying/expanding it like we do with our other example code.

### CODE REVIEW

**Comment by silby** [comment]
I can both `foo` _and_ `bar` successfully!

File: README.md [hunk 1]
```diff
@@ -1,37 +1,68 @@
...
+Deployment
+----------
 
-Development
------------
+Tag and push the container to the GCP Container Registry and Fux will auto deploy it.
 
-[Development instructions](DEVELOPMENT)  
+    export GIT_REPO_NAME=example-flask-app
+    gcloud --project uwit-mci-ist builds submit --tag gcr.io/uwit-mci-ist-reg/$GIT_REPO_NAME:$(git describe --always --tag) .
```
[the remark this answers is not in the document]
Comment (silby) [reply README.md]: (If this is the same basic command no matter the project I'd lean towards having a standard Makefile that does this on `make deploy` while automatic deployment from GitHub tags cooks)

File: docker-compose.yaml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+          listen 8001;
+          location /shared/ {
+              proxy_pass http://web-shared.default/shared/;
+          }
+          location /example-app/ {
+              proxy_set_header X-Forwarded-User test-user@washington.edu;
+              proxy_pass http://app:8000/example-app/;
+          }
+        }
+    command: /bin/sh -c "echo $$NGINX_CONFIG > /etc/nginx/conf.d/app.conf; nginx -g 'daemon off;'"
```
Comment (silby) [inline docker-compose.yaml]: example configs should self-document in comments. What does the nginx block here do? Can every application we run in dev use the same ports?

File: docker-compose.yaml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+  web-shared.default:
+    image: gcr.io/uwit-mci-ist-reg/web-shared:latest
+
+  redis:
+    image: "redis:5-alpine"
```
Comment (silby) [inline docker-compose.yaml:46]: What's this for?

File: Dockerfile [hunk 4]
```diff
@@ -41,4 +41,4 @@ EXPOSE 8000
 ENTRYPOINT ["/home/toolop/venv/bin/gunicorn", "flask_app.loader:app"]
 
 # default parameters that can be overridden with: docker run <image> new params
-CMD ["-b", ":8000", "--worker-class=eventlet", "--access-logfile", "-", "--log-level", "INFO"]
+CMD ["-b", ":8000", "--access-logfile", "-", "--log-level", "INFO"]
```
Comment (silby) [inline Dockerfile]: commit message doesn't explain this change, belongs in the next commit.

File: docker-compose.yaml [hunk 5]
```diff
@@ -44,7 +44,7 @@ services:
     command: /bin/sh -c "echo $$NGINX_CONFIG > /etc/nginx/conf.d/app.conf && exec nginx -g 'daemon off;'"
 
   web-shared.default:
-    image: gcr.io/uwit-mci-ist-reg/web-shared:latest
+    image: gcr.io/uwit-mci-ist/web-shared:latest
```
Comment (silby) [inline docker-compose.yaml]: this belongs in a different commit.

File: docker-compose.yaml [hunk 6]
```diff
@@ -19,7 +19,7 @@ services:
     # for local development mount and run your local copy instead of the one in the container
     volumes:
     - ./flask_app:/home/toolop/flask_app
-    entrypoint: /home/toolop/venv/bin/gunicorn flask_app.loader:app -b :8000 --worker-class=eventlet --access-logfile - --log-level INFO --reload
+    entrypoint: /home/toolop/venv/bin/gunicorn flask_app.loader:app -b :8000 --workers 3 --access-logfile - --log-level INFO --reload
```
Comment (silby) [inline docker-compose.yaml]: why change? (I don't object but let's explain it somewhere)

File: Dockerfile [hunk 7]
```diff
@@ -13,14 +13,20 @@ COPY . app
...
-WORKDIR /home/toolop/app
-RUN python setup.py test
+RUN python setup.py -q test
 
-# install the app
-RUN pip install .
+# install the app to the venv
+RUN pip install -q .
 
+# show what can be put in requirments.txt to pin dependencies for future container builds
```
Comment (silby) [inline Dockerfile:28]: `s/requirments/requirements`

File: README.md [hunk 8]
```diff
@@ -27,9 +27,22 @@ Testing
 The same container that runs locally for testing should run in production.
 Configuration and secrets are provided to the container at runtime.
 
-Unit tests are run every time a container is built
 
+Reproducable container building
```
Comment (silby) [inline README.md:29]: Use `### subheaders`

File: setup.py [hunk 9]
```diff
@@ -13,7 +13,7 @@ def readme():
     packages=find_packages(),
     include_package_data=True,
     zip_safe=False,
-    install_requires=['flask', 'gunicorn', 'eventlet', 'psycopg2-binary', 'requests', 'redis', 'flask-caching', 'Flask-Assets'],
+    install_requires=['flask', 'gunicorn', 'flask-caching', 'redis', 'whitenoise', 'psycopg2-binary', 'requests'],
```
Comment (silby) [inline setup.py]: `eventlet` part of this commit belongs elsewhere

File: docker-compose.yaml [hunk 10]
```diff
@@ -12,40 +12,82 @@ services:
...
 
   redis:
     image: "redis:5-alpine"
+
+
+  web-shared.default:
+    image: gcr.io/uwit-mci-ist/web-shared:latest
+
+
+  whoami.default:
```
Comment (silby) [inline docker-compose.yaml:53]: explain what this does/is for.

**Comment by silby** [comment]
Also I suggest removing your `DEVELOPMENT` notes from version control and using them as the start of a wiki page.

**Comment by benroy73** [comment]
Those are good ideas. I'll look into modifying the commit history, I haven't done that before but this seems like a good place to have that be nicer.

**Comment by benroy73** [comment]
Look like it will be easier if I just close this PR and rebase everything and try to make the commit messages better.

**Comment by silby** [comment]
whatever's easiest, but if you rebase your branch locally and then force-push it to github (`git push origin +convert-to-mci-container` the PR will update with the new commits.
